### PR TITLE
feat: Allow minting ERC20TransferableReceivables on behalf of other users

### PR DESCRIPTION
**Pull request by mliu**
This PR improves the ERC20TransferableReceivable feature by allowing anyone to mint an NFT on behalf of the payee. Currently, only the initial payee can mint their own NFT, logic we introduced to prevent a frontrunning attack. However, this process can be cumbersome for users who have to manually mint multiple requests at once.

The proposed solution is to modify the mint call to accept an address payee, and set the initial ERC721 owner as the payee. This change would make the minting process more flexible, allowing others to cover the minting costs, such as Request finance in the case of automatically recurring invoices.

To maintain security, the existing system of allowing multiple minted receivables for an invoice, keyed by a hash of the initialRecipient and the paymentReference, would still be used. This prevents anyone from frontrunning the call and setting themselves as the initialRecipient since the hash would still be invalid.

### CODE REVIEW

File: packages/smart-contracts/src/contracts/ERC20TransferableReceivable.sol [hunk 1]
```diff
@@ -5,29 +5,60 @@ import '@openzeppelin/contracts/utils/Counters.sol';
...
   Counters.Counter private _paymentId;
 
-  // Counter for uniquely identifying receivable tokens
+  /**
+   * @dev Counter for uniquely identifying receivables
+   */
   Counters.Counter private _receivableTokenId;
 
+  /**
+   * @dev Struct for storing receivable information
```
[the remark this answers is not in the document]
Comment (mliu) [reply packages/smart-contracts/src/contracts/ERC20TransferableReceivable.sol:28]: will change

File: packages/smart-contracts/src/contracts/ERC20TransferableReceivable.sol [hunk 2]
```diff
@@ -58,6 +102,16 @@ contract ERC20TransferableReceivable is ERC721, ERC721Enumerable, ERC721URIStora
     paymentProxy = _paymentProxyAddress;
   }
 
+  /**
+   * @notice Pay the owner of the specified receivable with the provided amount of ERC20 tokens.
+   * @param receivableTokenId The ID of the receivable token to pay.
+   * @param amount The amount of ERC20 tokens to pay the owner.
+   * @param paymentReference A reference for the payment.
+   * @param feeAmount The amount of ERC20 tokens to be paid as a fee for the transaction.
```
[the remark this answers is not in the document]
Comment (mliu) [reply packages/smart-contracts/src/contracts/ERC20TransferableReceivable.sol:110]: will change

File: packages/smart-contracts/src/contracts/ERC20TransferableReceivable.sol [hunk 3]
```diff
@@ -98,7 +152,17 @@ contract ERC20TransferableReceivable is ERC721, ERC721Enumerable, ERC721URIStora
     );
   }
 
+  /**
+   * @notice Mint a new transferable receivable.
+   * @param owner The address of the owner of the receivable token to be minted.
+   * @param paymentReference A reference for the payment.
+   * @param amount The amount of ERC20 tokens to be paid.
+   * @param erc20Addr The address of the ERC20 token to be used as payment.
+   * @param newTokenURI The URI of the token.
```
[the remark this answers is not in the document]
Comment (mliu) [reply packages/smart-contracts/src/contracts/ERC20TransferableReceivable.sol:161]: will change

File: packages/smart-contracts/src/contracts/ERC20TransferableReceivable.sol [hunk 4]
```diff
@@ -98,7 +152,17 @@ contract ERC20TransferableReceivable is ERC721, ERC721Enumerable, ERC721URIStora
     );
   }
 
+  /**
+   * @notice Mint a new transferable receivable.
+   * @param owner The address of the owner of the receivable token to be minted.
```
[the remark this answers is not in the document]
Comment (mliu) [reply packages/smart-contracts/src/contracts/ERC20TransferableReceivable.sol:157]: will change

File: packages/smart-contracts/src/contracts/ERC20TransferableReceivable.sol [hunk 5]
```diff
@@ -121,10 +185,15 @@ contract ERC20TransferableReceivable is ERC721, ERC721Enumerable, ERC721URIStora
...
     });
 
-    _mint(msg.sender, currentReceivableTokenId);
+    _mint(owner, currentReceivableTokenId);
     _setTokenURI(currentReceivableTokenId, newTokenURI);
   }
 
+  /**
+   * @notice Get an array of all receivable token IDs owned by a specific address.
+   * @param _owner The address of the owner to retrieve the receivable tokens for.
```
[the remark this answers is not in the document]
Comment (mliu) [reply packages/smart-contracts/src/contracts/ERC20TransferableReceivable.sol:194]: will change

File: packages/smart-contracts/test/contracts/ERC20TransferableReceivable.test.ts [hunk 6]
```diff
@@ -48,35 +48,37 @@ describe('contract: ERC20TransferableReceivable', () => {
 
   describe('mint', async function () {
     it('revert with empty paymentReference', async function () {
```
[the remark this answers is not in the document]
Comment (mliu) [reply packages/smart-contracts/test/contracts/ERC20TransferableReceivable.test.ts:50]: will do

File: packages/smart-contracts/test/contracts/ERC20TransferableReceivable.test.ts [hunk 7]
```diff
@@ -183,7 +185,7 @@ describe('contract: ERC20TransferableReceivable', () => {
     });
 
     it('allow multiple mints per receivable', async function () {
-      await receivable.connect(user2).mint(paymentRef, amount, testToken.address, '1');
+      await receivable.connect(user2).mint(user2Addr, paymentRef, amount, testToken.address, '1');
       const key = ethers.utils.solidityKeccak256(['address', 'bytes'], [user1Addr, paymentRef]);
       expect(await receivable.receivableTokenIdMapping(key)).to.equals(tokenId);
     });
```
[the remark this answers is not in the document]
Comment (mliu) [reply packages/smart-contracts/test/contracts/ERC20TransferableReceivable.test.ts:191]: Yeah sure I can add a test here. We also have a test for this in the payment-processor package: https://github.com/00labs/requestNetwork/blob/7c01bd329b3b754b115861f1301fa6a50cba4ccd/packages/payment-processor/test/payment/erc20-transferable-receivable.test.ts#L249

**Review by MantisClone — APPROVED** [review]

**Pull request merged**